sources/healthdog/src/service_check.rs

```diff
@@ -0,0 +1,127 @@
+use crate::error::{self, Result};
+use lazy_static::lazy_static;
+use log::trace;
+use regex::Regex;
+use snafu::ResultExt;
+use std::process::Command;
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) struct ServiceHealth {
+    /// Whether or not the service is healthy.
+    pub(crate) is_healthy: bool,
+    /// In the event of an unhealthy service, the service's exit code (if found).
+    pub(crate) exit_code: Option<i32>,
+}
+
+pub(crate) trait ServiceCheck {
+    /// Checks the given service to see if it is healthy.
+    fn check(&self, service_name: &str) -> Result<ServiceHealth>;
+}
+
+pub(crate) struct SystemdCheck {}
+
+impl ServiceCheck for SystemdCheck {
+    fn check(&self, service_name: &str) -> Result<ServiceHealth> {
+        if is_ok(service_name)? {
+            return Ok(ServiceHealth {
+                is_healthy: true,
+                exit_code: None,
+            });
+        }
+        Ok(ServiceHealth {
+            is_healthy: false,
+            exit_code: parse_service_exit_code(service_name)?,
+        })
+    }
+}
+
+struct Outcome {
+    exit: i32,
+    stdout: String,
+}
+
+impl Outcome {
+    fn is_exit_true(&self) -> bool {
+        self.exit == 0
+    }
+}
+
+fn systemctl(args: &[&str]) -> Result<Outcome> {
+    trace!("calling systemctl with '{:?}'", args);
+    let output = Command::new("systemctl")
+        .args(args)
+        .output()
+        .context(error::Command {
+            command: "systemctl",
+            args: args.iter().map(|&s| s.to_owned()).collect::<Vec<String>>(),
+        })?;
+    Ok(Outcome {
+        exit: output.status.code().unwrap_or(-1),
+        stdout: String::from_utf8_lossy(output.stdout.as_slice()).into(),
+    })
+}
+
+fn is_active(service: &str) -> Result<bool> {
+    let outcome = systemctl(&["is-active", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_failed(service: &str) -> Result<bool> {
+    let outcome = systemctl(&["is-failed", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_ok(service: &str) -> Result<bool> {
+    Ok(!is_failed(service)? && is_active(service)?)
```

Comment:
Is is-active always safe?  What if we wanted to check that a oneshot service was successful?

Reply:
I think I agree with Jamie on this:

> It'd be nice to eventually add support for non-lingering oneshot services, but I think this is a good place to start.

The benefit of checking is-active is that we would know if a service never started (as opposed to started and failed). For now, supporting a oneshot service with RemainAfterExit=false seems like YAGNI. But if we do, I can imagine ways of adding supporting for it in a backwards compatible way.